_posts/2017-09-15-monitoring-your-microservice-with-micrometer.md

```diff
@@ -0,0 +1,112 @@
+tags: [microservices, monitoring]
+category: Microservices
+comments: true
+---
+# Monitoring your microservices using Micrometer.io
+
+When we want to instrument our application, we don't want to worry about which monitoring system we want to use now or in the future.
+Nor do we want to change a lot of code throughout our microservice because we need to change from system X to system Y.
+
+## Meet Micrometer!
+
+So what is Micrometer you ask?
+
+Basically, it comes down to this line:
+> Think SLF4J, but for metrics.
+
+[Micrometer](https://micrometer.io) provides a simple facade over the instrumentation clients for the most popular monitoring systems.
+It allows you to instrument your code with dimensional metrics with a vendor-neutral interface and decide on the monitoring system as a last step.
+Using this interface, we can support multiple monitoring systems and switch easily to an other system with little to no hassle.
+It already contains built-in support for [**Prometheus**](https://prometheus.io), Netflix [**Atlas**](https://github.com/Netflix/atlas), and [**Datadog**](https://www.datadoghq.com), while [InfluxDB](https://www.influxdata.com), [statsd](https://github.com/etsy/statsd), and [Graphite](https://graphiteapp.org) are on their way!
+
+
+## Using Micrometer in your application
+
+Starting with Spring Boot 2, Micrometer becomes the defacto instrumentation library that will be powering the delivery of application metrics from Spring.
+Luckily for us, they also backported this functionality to Spring Boot 1.x through an additional library dependency!
+Just add the `micrometer-spring-legacy` module together with the additional monitoring system module, and you're good to go!
+
+In Gradle:
+
+``` gradle
+compile 'io.micrometer:micrometer-spring-legacy:latest.release'
+```
+
+Or in Maven:
+
+``` maven
+<dependency>
+  <groupId>io.micrometer</groupId>
+  <artifactId>micrometer-spring-legacy</artifactId>
+  <version>${micrometer.version}</version>
+</dependency>
+```
+
+## Creating metrics
+
+There are a couple of ways to create meters.
+We will cover all different types, when to use them, and furthermore how to implement them.
+
+### Dimensions/Tags
+
+A meter is uniquely identified by its name and dimensions (also called tags).
+Dimensions are a way of adding dimensions to metrics, so they can be sliced, diced, aggregated and compared.
+For example, we have a meter named `http.requests` with a tag `uri`.
+With this meter we could see the overall amount of HTTP request, but also have the option to drill down and see the amount of HTTP requests for a specific URI.
+
+### Counters
+
+Counters are a cumulative metric that represents a single numerical value that only ever goes up.
+They are typically used to count requests served, tasks completed, errors occurred, etc.
+Counters should not be used to expose current counts of items whose number can also go down, gauges are a better fit for this use case.
```

Comment:
make this last line bold? seems important

Reply:
I changed it to "Counters should **not** be used to expose current counts"